keps/sig-cli/2906-kustomize-function-catalog/README.md

```diff
@@ -0,0 +1,667 @@
+
+As an example, the Helm functionality built into Kustomize currently relies on the user to install Helm as a separate step. This in turn requires the use of special flags on invocation to use, for security reasons. If the Helm integration was instead built and distributed as a container based plugin, the implementation could instead use the Helm Go packages and be built independently of Kustomize and distributed to users of Kustomize through an official channel. 
+
+### Goals
+
+Develop an API (`kind`) for Kustomize that is focused on plugin provider discovery, as well as guidelines and recommendations for distribution of plugin providers and associated resources, such as schema definitions.
+
+A successful implementation of this API should have the following characteristics:
+
+1. Plugin based workflows are driven by seamless invocation of sets of plugin providers without individual out-of-band discovery or installation steps for specific providers.
+1. The new API is integrated with the existing Kustomize tool (i.e. `kustomize build`) through references provided in Kustomization, Component, and Composition resources.
+1. Eligible Kustomize functionality could be extracted and distributed as official extensions. This won't be completed as part of this KEP but the required changes to support this will be implemented. 
+
+### Non-Goals
+
+1. Support anything other than KRM-style plugins that follow the [functions spec](https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md)
+1. Directly implement capabilities to publish plugin providers or other resources to OCI registries 
+
+## Proposal
+
+In order to standardize plugin discovery, introduce a `Catalog` API `kind` recognized by `kustomize build`. One or more `Catalog` resources can be referenced from any Kustomize kind as either a local file or remote reference available as from an HTTP(s) endpoint or as an [OCI artifact](https://github.com/opencontainers/artifacts).
+
+A `Catalog` will contain a collection of one or more modules that can be used with a Kustomize resource. 
+
+A minimal example is shown below:
+
+```yaml
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata: 
+  name: "example-co-plugins"
+spec: 
+  modules: 
+  - apiVersion: example.com/v1
+    kind: JavaApplication
+    description: "A Kustomize plugin provider that can handle Java apps"
+    provider: 
+      container: 
+        image: example/module_providers/java:v1.0.0
+```
+
+This will enable a users to use plugin providers with provider information omitted, such as:
+
+```yaml
+# app/kustomization.yaml
+kind: Kustomization
+catalogs:
+  - https://example.com/plugins/catalog.yaml
```

Comment:
it is an option, I fear that that functionality has a larger discussion but could be used to accomplish all of the remote resource pulling.

Reply:
I think that it would be desirable to support both local and remote references, although we could further restrict this to _only_ remote references if absolutely required. I've updated the wording to specify that both are allowed, happy to iterate further on the text. 

I do like the `kustomize localize` suggestion, I think that would be an interesting way to handle it. 

I think that a versioned OCI artifact would be a good (perhaps ideal) way to reference these in an immutable way, but it would probably be best to allow some flexibility for end users. We could restrict this to Local,  Git, or OCI based references?